### PR TITLE
DirectAllocator: reduce the amount of redundant memcpy calls on Windows

### PR DESCRIPTION
**Pull request by squeek502**
This is a follow-up to #2312

Previously the memory would be copied to a different aligned address in some cases where the old offset could have been used. This fixes it so that it will always try to use the old offset when possible, and only uses a different offset if the old one is truly invalid (not aligned or not enough space to store the alloc at the old offset).

Also changed the realloc memory copy calls to use `std.math.min` as mentioned here: https://github.com/ziglang/zig/pull/2312#issuecomment-486509294

### CODE REVIEW

File: std/heap.zig [hunk 1]
```diff
@@ -170,16 +166,20 @@ pub const DirectAllocator = struct {
...
                 const new_root_addr = @ptrToInt(new_ptr);
-                const adjusted_addr = new_root_addr + offset;
-                const new_adjusted_addr = mem.alignForward(new_root_addr, new_align);
-                // If HeapReAlloc didn't happen to move the memory to the new alignment
-                // then we need to copy it
-                if (new_adjusted_addr != adjusted_addr) {
+                var new_adjusted_addr = new_root_addr + offset;
+                const offset_is_valid = new_adjusted_addr + new_size + @sizeOf(usize) <= new_root_addr + amt;
+                const offset_is_aligned = new_adjusted_addr % new_align == 0;
+                if (!offset_is_valid or !offset_is_aligned) {
```
Comment (squeek502) [inline std/heap.zig]: I'm not confident in the style of this conditional. Suggestions for how to improve it are welcome.

Comment (andrewrk) [reply std/heap.zig]: It looks fine to me - what are you concerned about?

Comment (squeek502) [reply std/heap.zig]: If it looks fine to you then there's no issue. Wasn't sure about pulling out the conditions into variables.

Comment (andrewrk) [reply std/heap.zig]: Yep, in general I think that's a great technique for clarity.

**Pull request merged**